megamek/src/megamek/common/EquipmentType.java

```diff
@@ -553,10 +555,16 @@ public boolean hasModeType(String modeType) {
     		return false;
     	}
     	
-    	for(EquipmentMode mode : modes) {
-    		if(mode.getName().equals(modeType)) {
-    			return true;
-    		}
+    	// sometimes, the list of modes may be modified while we're looping through it
+    	// so let's do this in a thread-safe way
+    	List<EquipmentMode> threadSafeModes = Collections.synchronizedList(modes);
```

Comment:
`Collections.synchronizedList` protects the given methods on `List` from concurrent access, however, they do not protect iteration. So you'd need to use `synchronized`.

Except, `threadSafeModes` isn't being used for all of your mutable uses of `modes` (`add`, `remove`). As it stands, you'd likely be better served by simply wrapping every usage of `modes` in a `synchronized` block on `modes` itself.

Reply:
Also it looks like `modes` is a `Vector`, which already has the same behavior as a `synchronizedList`. So basically, either `addMode` or `removeMode` was called while this was going on. My guess is a network delay or GC pause gave it enough timing to fail.